sdks/java/core/src/main/java/org/apache/beam/sdk/coders/GeneratedRowCoder.java

```diff
@@ -0,0 +1,316 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.coders;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import net.bytebuddy.ByteBuddy;
+import net.bytebuddy.description.modifier.FieldManifestation;
+import net.bytebuddy.description.modifier.Ownership;
+import net.bytebuddy.description.modifier.Visibility;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.description.type.TypeDescription.ForLoadedType;
+import net.bytebuddy.dynamic.DynamicType;
+import net.bytebuddy.dynamic.loading.ClassLoadingStrategy;
+import net.bytebuddy.dynamic.scaffold.InstrumentedType;
+import net.bytebuddy.implementation.FixedValue;
+import net.bytebuddy.implementation.Implementation;
+import net.bytebuddy.implementation.bytecode.ByteCodeAppender;
+import net.bytebuddy.implementation.bytecode.Duplication;
+import net.bytebuddy.implementation.bytecode.StackManipulation;
+import net.bytebuddy.implementation.bytecode.StackManipulation.Compound;
+import net.bytebuddy.implementation.bytecode.TypeCreation;
+import net.bytebuddy.implementation.bytecode.assign.TypeCasting;
+import net.bytebuddy.implementation.bytecode.collection.ArrayFactory;
+import net.bytebuddy.implementation.bytecode.member.FieldAccess;
+import net.bytebuddy.implementation.bytecode.member.MethodInvocation;
+import net.bytebuddy.implementation.bytecode.member.MethodReturn;
+import net.bytebuddy.implementation.bytecode.member.MethodVariableAccess;
+import net.bytebuddy.matcher.ElementMatchers;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.Schema.TypeName;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.reflect.FieldValueGetter;
+
+/**
+ * A utility for automatically generating a {@link Coder} for {@link Row} objects corresponding to
+ * a specific schema. The resulting coder is loaded into the default ClassLoader and returned.
+ *
+ */
+public abstract class GeneratedRowCoder {
```

Comment:
The more I think about this class, the more I think `GeneratedRowCoder` is somewhat misguiding name. For me, this class is more like `RowCoderGenerator`, and the actual generated coder is a Coder<Row> class that looks something like:
```
class Coder<Row> {
  Coder[] FIELD_CODERS;
  Schema getSchema() {};
  encode(Row value, OutputStream outputStream) {encodeDeleagate(FIELD_CODERS, Row value, OutputStream outputStream)};
  encodeDelegate(FIELD_CODERS, Row value, OutputStream outputStream) {}
  .... // decoder code
}
```

I suggest renaming this class to RowCoderGenerator, and add a high level javadoc explaining how a generated row coder object looks like. Knowing what object is being generated is a huge help understanding the byte buddy code.

Reply:
Or we can have a GeneratedRowCoder interface too.